File: src/main/java/com/github/morningzeng/toolset/ui/crypto/SM4Component.java
package com.github.morningzeng.toolset.ui.crypto;

import cn.hutool.core.util.HexUtil;
import cn.hutool.crypto.Mode;
import cn.hutool.crypto.Padding;
import cn.hutool.crypto.symmetric.SM4;
import com.github.morningzeng.toolset.model.SymmetricCryptoProp;
import com.github.morningzeng.toolset.utils.GridBagUtils.Row;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.ComboBox;

/**
 * @author Morning Zeng
 * @since 2024-10-31
 */
public final class SM4Component extends AbstractSymmetricCryptoComponent {

    private final ComboBox<Mode> modeComboBox = new ComboBox<>(Mode.values());
    private final ComboBox<Padding> paddingComboBox = new ComboBox<>(Padding.values());

    public SM4Component(final Project project) {
        super(project);
        this.initLayout();
        this.initAction();
    }

    @Override
    void cryptoRow(final Row<AbstractSymmetricCryptoComponent> row) {
        row.newCell().add(this.modeComboBox)
                .newCell().add(this.paddingComboBox);
    }

    @Override
    String enc(final SymmetricCryptoProp cryptoProp) {
        return this.sm4(cryptoProp).encryptHex(this.decryptArea.getText());
    }

    @Override
    String dec(final SymmetricCryptoProp cryptoProp) {
        return this.sm4(cryptoProp).decryptStr(this.encryptArea.getText());
    }

    private SM4 sm4(final SymmetricCryptoProp cryptoProp) {
        final Mode mode = this.modeComboBox.getItem();
        final Padding padding = this.paddingComboBox.getItem();
        final byte[] key = HexUtil.decodeHex(cryptoProp.getKey());
        final byte[] iv = HexUtil.decodeHex(cryptoProp.getIv());
        return new SM4(mode, padding, key, iv);
    }
}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/view/crypto/SM4Component.kt
package com.github.morningzeng.toolbox.ui.view.crypto

import cn.hutool.core.util.HexUtil
import cn.hutool.crypto.Mode
import cn.hutool.crypto.Padding
import cn.hutool.crypto.symmetric.SM4
import com.github.morningzeng.toolbox.enums.DataToBinaryTypeEnum
import com.github.morningzeng.toolbox.model.CryptoSymmetric
import com.github.morningzeng.toolbox.utils.GridBagUtils
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import javax.swing.JComponent

/**
 * @author Morning Zeng
 * @since 2025-05-22
 */
class SM4Component(
    project: Project
) : AbstractCryptoSymmetricComponent(project) {

    private val modeComboBox: ComboBox<Mode> = ComboBox(Mode.entries.toTypedArray())
    private val paddingComboBox: ComboBox<Padding> = ComboBox(Padding.entries.toTypedArray())

    init {
        initLayout()
        initAction()
    }

    override fun cryptoRow(row: GridBagUtils.Row<out JComponent>) {
        GridBagUtils.builder()
            .row {
                it.fill(GridBagUtils.GridBagFill.HORIZONTAL)
                    .cell().add(modeComboBox)
                    .cell().add(paddingComboBox)
            }
            .build()
            .apply { row.cell().add(this) }
    }

    override fun encrypt(prop: CryptoSymmetric): String {
        return when (decryptedContentComboBox.component.item) {
            DataToBinaryTypeEnum.HEX -> sm4(prop).encryptHex(decryptArea.text)
            DataToBinaryTypeEnum.BASE64 -> sm4(prop).encryptBase64(decryptArea.text)
            DataToBinaryTypeEnum.TEXT -> String(sm4(prop).encrypt(decryptArea.text.toByteArray()), Charsets.UTF_8)
        }
    }

    override fun decrypt(prop: CryptoSymmetric): String {
        return sm4(prop).decryptStr(encryptArea.text)
    }

    private fun sm4(prop: CryptoSymmetric): SM4 {
        val mode = modeComboBox.item
        val padding = paddingComboBox.item
        val key = HexUtil.decodeHex(prop.key)
        val iv = HexUtil.decodeHex(prop.iv)
        return SM4(mode, padding, key, iv)
    }

}